diff -bur ndiswrapper-1.7.orig/driver/loader.c ndiswrapper-1.7/driver/loader.c
--- ndiswrapper-1.7.orig/driver/loader.c	2006-10-14 10:44:02.000000000 +0200
+++ ndiswrapper-1.7/driver/loader.c	2006-10-14 10:44:52.000000000 +0200
@@ -766,7 +766,9 @@
 #ifdef CONFIG_USB
 	if (wrap_usb_devices) {
 		memset(&wrap_usb_driver, 0, sizeof(wrap_usb_driver));
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,16)
 		wrap_usb_driver.owner = THIS_MODULE;
+#endif
 		wrap_usb_driver.name = DRIVER_NAME;
 		wrap_usb_driver.id_table = wrap_usb_devices;
 		wrap_usb_driver.probe = wrap_pnp_start_usb_device;
